lib/typecheck.ts

```diff
@@ -679,6 +679,61 @@ export default class TypeChecker {
         return Promise.resolve();
     }
 
+    private async _typeCheckJoin(ast : Ast.JoinExpression, scope : Scope) {
+        assert(ast.lhs.schema && ast.rhs.schema);
+        const name = `join(${ast.lhs.schema.name},${ast.rhs.schema.name})`;
+        if (!ast.lhs.schema.is_list || !ast.rhs.schema.is_list)
+            throw new TypeError('Both sides of a join must be a list query');
+        const classDef = null; // FIXME: what should this be for a join expression? 
+        const qualifiers = {
+            is_list : true,
+            is_monitorable: ast.lhs.schema.is_monitorable || ast.rhs.schema.is_monitorable
+        };
+        const args = [];
+        for (const arg of ast.lhs.schema.iterateArguments()) {
+            if (arg.is_input)
+                continue;
+            const newArg = arg.clone();
+            newArg.name = `first.${arg.name}`;
+            args.push(newArg);
+        }
+        for (const arg of ast.rhs.schema.iterateArguments()) {
+            if (arg.is_input)
+                continue;
+            const newArg = arg.clone();
+            newArg.name = `second.${arg.name}`;
+            args.push(newArg);
+        }
+        ast.schema = new Ast.FunctionDef(null, 'query', classDef, name, [], qualifiers, args);
+        if (ast.condition) 
+            await this._typeCheckJoinCondition(ast.condition, ast.schema);
+        return Promise.resolve();
+    }
+
+    private async _typeCheckJoinCondition(ast : Ast.BooleanExpression, schema : Ast.FunctionDef) {
+        if (ast instanceof Ast.AtomBooleanExpression) {
+            if (!ast.name.startsWith('first.'))
+                throw new TypeError('Fields on the left-hand side in a join condition needs a `first.` prefix');
+            if (!schema.hasArgument(ast.name))
+                throw new TypeError(`Field ${ast.name.slice('first.'.length)} is not available on the left-hand side of the join`);
+            if (!(ast.value instanceof Ast.VarRefValue))
+                throw new TypeError(`Only field reference can be used in join condition`);
+            if (!ast.value.name.startsWith('second.'))
+                throw new TypeError('Fields on the right-hand side in a join condition needs a `second.` prefix');
+            if (!schema.hasArgument(ast.value.name))
+                throw new TypeError(`Field ${ast.value.name.slice('second.'.length)} is not available on the left-hand side of the join`);
+            const type_lhs = schema.getArgType(ast.name);
+            const type_rhs = schema.getArgType(ast.value.name);
+            ast.overload = await this._resolveFilterOverload(type_lhs!, ast.operator, type_rhs!);
+        } else if (ast instanceof Ast.NotBooleanExpression) {
+            this._typeCheckJoinCondition(ast.expr, schema);
+        } else if (ast instanceof Ast.AndBooleanExpression || ast instanceof Ast.OrBooleanExpression) {
+            ast.operands.forEach((o) => this._typeCheckJoinCondition(o, schema));
+        } else {
```

Comment:
True and False?

Reply:
Technically, it's impossible if all lhs is a field from one table, and rhs is a field from another table...

Reply:
But given that you suggest to use True/False boolean expression instead of `null` for the condition, it makes sense to add them here

Reply:
fixed in 61a6e46.